Compilatoin failure with gcc 5.4.0

```
/home/wthrowe/spectre/tests/Unit/Utilities/Test_ConstantExpressions.cpp:83:1: error: non-constant condition for static assertion
 static_assert(array_equal(array1, array1_copy), "Failed testing array_equal");
 ^
In file included from /home/wthrowe/spectre/tests/Unit/Utilities/Test_ConstantExpressions.cpp:6:0:
/home/wthrowe/spectre/tests/Unit/Utilities/Test_ConstantExpressions.cpp:83:26:   in constexpr expansion of ‘array_equal<double, double, 3ul>({anonymous}::array1, {anonymous}::array1_copy, 0ul)’
/home/wthrowe/spectre/src/Utilities/ConstantExpressions.hpp:216:29: error: ‘constexpr const typename Cont::value_type& gsl::at(const Cont&, Size) [with Cont = std::array<double, 3ul>; Size = long unsigned int; typename Cont::value_type = double]’ called in a constant expression
   return i < size ? (gsl::at(lhs, i) == gsl::at(rhs, i) and
                             ^
In file included from /home/wthrowe/spectre/src/Utilities/ConstantExpressions.hpp:12:0,
                 from /home/wthrowe/spectre/tests/Unit/Utilities/Test_ConstantExpressions.cpp:6:
/home/wthrowe/spectre/src/Utilities/Gsl.hpp:119:66: note: ‘constexpr const typename Cont::value_type& gsl::at(const Cont&, Size) [with Cont = std::array<double, 3ul>; Size = long unsigned int; typename Cont::value_type = double]’ is not usable as a constexpr function because:
 SPECTRE_ALWAYS_INLINE constexpr const typename Cont::value_type& at(
                                                                  ^
/home/wthrowe/spectre/tests/Unit/Utilities/Test_ConstantExpressions.cpp:84:1: error: non-constant condition for static assertion
 static_assert(not array_equal(array1, std::array<double, 3>{{2.2, 3, 4}}),
 ^
In file included from /home/wthrowe/spectre/tests/Unit/Utilities/Test_ConstantExpressions.cpp:6:0:
/home/wthrowe/spectre/tests/Unit/Utilities/Test_ConstantExpressions.cpp:84:30:   in constexpr expansion of ‘array_equal<double, double, 3ul>({anonymous}::array1, std::array<double, 3ul>{std::__array_traits<double, 3ul>::_Type{2.2000000000000002e+0, 3.0e+0, 4.0e+0}}, 0ul)’
/home/wthrowe/spectre/src/Utilities/ConstantExpressions.hpp:216:29: error: ‘constexpr const typename Cont::value_type& gsl::at(const Cont&, Size) [with Cont = std::array<double, 3ul>; Size = long unsigned int; typename Cont::value_type = double]’ called in a constant expression
   return i < size ? (gsl::at(lhs, i) == gsl::at(rhs, i) and
                             ^
/home/wthrowe/spectre/tests/Unit/Utilities/Test_ConstantExpressions.cpp:86:1: error: non-constant condition for static assertion
 static_assert(array_equal(replace_at<1>(array1, 5.0),
 ^
In file included from /home/wthrowe/spectre/tests/Unit/Utilities/Test_ConstantExpressions.cpp:6:0:
/home/wthrowe/spectre/tests/Unit/Utilities/Test_ConstantExpressions.cpp:86:52:   in constexpr expansion of ‘replace_at<1ul, double, 3ul>({anonymous}::array1, 5.0e+0)’
/home/wthrowe/spectre/src/Utilities/ConstantExpressions.hpp:204:54:   in constexpr expansion of ‘ConstantExpression_detail::replace_at_helper<double, 3ul, {0ul}, {0ul}>((* & arr), (*(const double*)std::forward<double>(value)), 2ul, (std::make_integer_sequence<long unsigned int, 1ul>{}, std::make_integer_sequence<long unsigned int, 1ul>()), (std::make_integer_sequence<long unsigned int, 1ul>{}, std::make_integer_sequence<long unsigned int, 1ul>()))’
/home/wthrowe/spectre/src/Utilities/ConstantExpressions.hpp:195:33: error: ‘constexpr const typename Cont::value_type& gsl::at(const Cont&, Size) [with Cont = std::array<double, 3ul>; Size = long unsigned int; typename Cont::value_type = double]’ called in a constant expression
       {arr[I]..., value, gsl::at(arr, i + J)...}};
                                 ^
/home/wthrowe/spectre/tests/Unit/Utilities/Test_ConstantExpressions.cpp:91:1: error: non-constant condition for static assertion
 static_assert(
 ^
In file included from /home/wthrowe/spectre/tests/Unit/Utilities/Test_ConstantExpressions.cpp:6:0:
/home/wthrowe/spectre/tests/Unit/Utilities/Test_ConstantExpressions.cpp:92:16:   in constexpr expansion of ‘array_equal<long unsigned int, long unsigned int, 3ul>(std::array<long unsigned int, 3ul>{std::__array_traits<long unsigned int, 3ul>::_Type{1ul, 2ul, 5ul}}, make_array_from_list<brigand::list<brigand::integral_constant<long unsigned int, 1ul>, brigand::integral_constant<long unsigned int, 2ul>, brigand::integral_constant<long unsigned int, 5ul> >, 0u>(), 0ul)’
/home/wthrowe/spectre/src/Utilities/ConstantExpressions.hpp:216:29: error: ‘constexpr const typename Cont::value_type& gsl::at(const Cont&, Size) [with Cont = std::array<long unsigned int, 3ul>; Size = long unsigned int; typename Cont::value_type = long unsigned int]’ called in a constant expression
   return i < size ? (gsl::at(lhs, i) == gsl::at(rhs, i) and
                             ^
In file included from /home/wthrowe/spectre/src/Utilities/ConstantExpressions.hpp:12:0,
                 from /home/wthrowe/spectre/tests/Unit/Utilities/Test_ConstantExpressions.cpp:6:
/home/wthrowe/spectre/src/Utilities/Gsl.hpp:119:66: note: ‘constexpr const typename Cont::value_type& gsl::at(const Cont&, Size) [with Cont = std::array<long unsigned int, 3ul>; Size = long unsigned int; typename Cont::value_type = long unsigned int]’ is not usable as a constexpr function because:
 SPECTRE_ALWAYS_INLINE constexpr const typename Cont::value_type& at(
                                                                  ^
```
Experimenting, it doesn't like the use of strings and `Parallel::abort` in `Expects`.  I believe these are supposed to be allowed as long as the `Expects` condition is true so they are not evaluated.

--- a/src/Utilities/ConstantExpressions.hpp
+++ b/src/Utilities/ConstantExpressions.hpp
@@ -191,8 +191,10 @@ inline constexpr std::array<std::decay_t<T>, Size> replace_at_helper(
     const std::array<T, Size>& arr, const T& value, const size_t i,
     std::integer_sequence<size_t, I...> /*unused*/,
     std::integer_sequence<size_t, J...> /*unused*/) {
+  // clang-tidy: Cannot use gsl::at because we want constexpr evaluation and
+  // Parallel::abort violates this
   return std::array<std::decay_t<T>, Size>{
-      {arr[I]..., value, gsl::at(arr, i + J)...}};
+      {arr[I]..., value, arr[i + J]...}};  // NOLINT
 }
 }  // namespace ConstantExpression_detail
 
@@ -213,7 +215,9 @@ template <typename T, typename S, size_t size>
 inline constexpr bool array_equal(const std::array<T, size>& lhs,
                                   const std::array<S, size>& rhs,
                                   const size_t i = 0) noexcept {
-  return i < size ? (gsl::at(lhs, i) == gsl::at(rhs, i) and
-                     array_equal(lhs, rhs, i + 1))
+  // clang-tidy: Cannot use gsl::at because we want constexpr evaluation and
+  // Parallel::abort violates this
+  return i < size ? (lhs[i] == rhs[i]  // NOLINT
+                     and array_equal(lhs, rhs, i + 1))
                   : true;
 }

